File: Algorithm/2026/05/18/프로그래머스_코딩_테스트_공부/Solution_dp.java
import java.util.*;

class Solution {
    
    static final int INF = 305;
    
    public int solution(int alp, int cop, int[][] problems) {
        int[] targets = getTargets(problems);
        int targetA = targets[0];
        int targetC = targets[1];
        
        int[] startStats = clampInitialStats(alp, cop, targetA, targetC);
        alp = startStats[0];
        cop = startStats[1];
        
        return computeMinTime(alp, cop, targetA, targetC, problems);
    }
    
    private int[] getTargets(int[][] problems) {
        int targetA = 0;
        int targetC = 0;
        for (int[] prob : problems) {
            if (targetA < prob[0]) targetA = prob[0];
            if (targetC < prob[1]) targetC = prob[1];
        }
        return new int[]{targetA, targetC};
    }
    
    private int[] clampInitialStats(int alp, int cop, int targetA, int targetC) {
        if (alp > targetA) alp = targetA;
        if (cop > targetC) cop = targetC;
        return new int[]{alp, cop};
    }
    
    private int[][] createDpTable(int row, int col, int initAlp, int initCop) {
        int[][] dp = new int[row + 1][col + 1];
        for (int i = 0; i <= row; i++) {
            Arrays.fill(dp[i], INF);
        }
        dp[initAlp][initCop] = 0;
        return dp;
    }
    
    private int computeMinTime(int startA, int startC, int targetA, int targetC, int[][] problems) {
        int[][] dp = createDpTable(targetA, targetC, startA, startC);
        
        for (int i = startA; i <= targetA; i++) {
            for (int j = startC; j <= targetC; j++) {
                if (dp[i][j] == INF) continue;
                
                int nA = Math.min(i + 1, targetA);
                if (dp[nA][j] > dp[i][j] + 1) {
                    dp[nA][j] = dp[i][j] + 1;
                }
                
                int nC = Math.min(j + 1, targetC);
                if (dp[i][nC] > dp[i][j] + 1) {
                    dp[i][nC] = dp[i][j] + 1;
                }
                
                for (int[] prob : problems) {
                    if (prob[0] <= i && prob[1] <= j) {
                        int nextA = Math.min(i + prob[2], targetA);
                        int nextC = Math.min(j + prob[3], targetC);
                        int nextT = dp[i][j] + prob[4];
                        
                        if (dp[nextA][nextC] > nextT) {
                            dp[nextA][nextC] = nextT;
                        }
                    }
                }
            }
        }
        
        return dp[targetA][targetC];
    }
}

File: Algorithm/2026/05/18/프로그래머스_코딩_테스트_공부/Solution_dp.cpp
#include <string>
#include <vector>
#include <stdlib.h>
#define INF 305

using namespace std;

int solution(int alp, int cop, vector<vector<int>> problems) {
    int row = 0, col = 0, n = problems.size();
    for(int i = 0; i < n; i++) {
        if(row < problems[i][0]) row = problems[i][0];
        if(col < problems[i][1]) col = problems[i][1];
    }
    
    int** dp = (int**)malloc(sizeof(int*) * (row + 1));
    for(int i = 0; i <= row; i++) {
        dp[i] = (int*)malloc(sizeof(int) * (col + 1));
        for(int j = 0; j <= col; j++) {
            dp[i][j] = INF;
        }
    }
    if(alp > row) alp = row;
    if(cop > col) cop = col;
    dp[alp][cop] = 0;
    
    for(int i = alp; i <= row; i++) {
        for(int j = cop; j <= col; j++) {
            if(dp[i][j] == INF) continue;
            
            int nA = (i + 1) > row ? row : i + 1;
            if(dp[nA][j] > (dp[i][j] + 1)) {
                dp[nA][j] = dp[i][j] + 1;
            }
            
            int nC = (j + 1) > col ? col : j + 1;
            if(dp[i][nC] > (dp[i][j] + 1)) {
                dp[i][nC] = dp[i][j] + 1;
            }
            
            for(int k = 0; k < n; k++) {
                if(problems[k][0] <= i && problems[k][1] <= j) {
                    int nA_prob = (i + problems[k][2]) > row ? row : i + problems[k][2];
                    int nC_prob = (j + problems[k][3]) > col ? col : j + problems[k][3];
                    int nT_prob = dp[i][j] + problems[k][4];
                    
                    if(dp[nA_prob][nC_prob] > nT_prob) dp[nA_prob][nC_prob] = nT_prob;
                }
            }
        }
    }
    
    int answer = dp[row][col];
    for(int i = 0; i <= row; i++) {
        free(dp[i]);
    }
    free(dp);
    return answer;
}